Add well-known routes for Protected Audience API WPTs

orrb1 opened the pull request:
Web Platform Tests for the [Protected Audience API](https://github.com/WICG/turtledove/blob/main/FLEDGE.md) are in the process of being added. As part of this API, the browser issues a GET request to a well-known URI: `/.well-known/interest-group/permissions/`. This is described in more detail at
https://github.com/WICG/turtledove/blob/main/FLEDGE.md#13-permission-delegation.

In order to implement dynamic handling of these routes in WPT, we need to register these routes in the add_mount_point. Because this well-known URI ends in a '/', we also need special handling for directory-appearing PythonHandler-supported endpoints.

orrb1 committed:
Add well-known routes for Protected Audience API WPTs

Web Platform Tests for the [Protected Audience
API](https://github.com/WICG/turtledove/blob/main/FLEDGE.md) are in the
process of being added. As part of this API, the browser issues a GET
request to a well-known URI: `/.well-known/interest-group/permissions/`.
This is described in more detail at
https://github.com/WICG/turtledove/blob/main/FLEDGE.md#13-permission-delegation.

In order to implement dynamic handling of these routes in WPT, we need
to register these routes in the add_mount_point. Because this well-known
URI ends in a '/', we also need special handling for directory-appearing
PythonHandler-supported endpoints.

orrb1 committed:
Fix indentation in change to handlers.py

orrb1 committed:
Support default python files matching the pattern default*.py.

In order to support e.g. text substition, which requires a special
filename, e.g. default.sub.py, expand support for loading the default
file in a directory-appearing endpoint handled by a Python handler to
support any default file matching the pattern default*.py. In case
multiple files match this pattern, the first one found is used.

orrb1 committed:
Improved handling for default files matching default*.py.

orrb1 committed:
Improved handling for default files matching default*.py.

foolip reviewed (commented):
Can you add tests for the new functionality?

orrb1 committed:
Add tests for directory special casing for PythonHandler.

This also makes the directory special casing more deterministic by
explicitly selecting the first lexicographically sorted file match the
default*.py pattern in the requested path.

orrb1 commented on tools/wptserve/wptserve/handlers.py:
```diff
@@ -298,6 +302,12 @@ def _load_file(self, request, response, func):
 
+        # Find a default Python file if the specified path is a directory
+        if os.path.isdir(path):
+            default_py = next(pathlib.Path(path).glob("default*.py"), None)
```
Good point. I originally wrote this to assume it should just look for default.py, and only after that thought that test authors might need substitution or other pipe support. I had considered making the case where wptserve found multiple "default*.py" files an error instead of silently selecting one, but considered that there might be a reason why someone's handler might need to happen to match default*.py. I think having some deterministic behavior makes sense, and I've gone with lexicographical order as the most obvious choice. Thanks!

orrb1 commented:
> Can you add tests for the new functionality?

Absolutely. I've added tests for the new Python special case for directory paths. Thanks!

foolip reviewed (approved):
This all looks good to me. I see there's still a CI failure. I'm checking out for today, but if you find it's not related to these changes, feel free to poke me or @past for an admin merge.

orrb1 committed:
Changed AsIsHandler to throw HTTPException instead of AssertionError.

Also added a test that shows AsIsHandler failing when trying to process
a directory.

orrb1 commented:
> This all looks good to me. I see there's still a CI failure. I'm checking out for today, but if you find it's not related to these changes, feel free to poke me or @past for an admin merge.

Thanks for taking a look! I've seen these CI errors before - they seem to be tooling errors - and another git push resolved them. In this case, it gave me an excuse to revisit the change I made to the AsIsHandler - which was originally an assert that the path isn't a directory, and now returns a 500 if it encounters that - and added a test for that. Thanks again.
